File: frontend/src/services/noteService.ts
import api from "./api";
import type { Note } from "@/types";

export const noteService = {
  async getAll(params?: { subject_id?: string; page?: number; limit?: number }) {
    const response = await api.get<{ notes: Note[]; total: number }>("/api/notes", { params });
    return response.data;
  },

  async getById(id: string) {
    const response = await api.get<Note>(`/api/notes/${id}`);
    return response.data;
  },

  async create(data: { title: string; content?: string; is_markdown?: boolean; subject_id?: string; chapter_id?: string }) {
    const response = await api.post<Note>("/api/notes", data);
    return response.data;
  },

  async update(id: string, data: Partial<Note>) {
    const response = await api.put<Note>(`/api/notes/${id}`, data);
    return response.data;
  },

  async delete(id: string) {
    await api.delete(`/api/notes/${id}`);
  },
};

File: backend/app/services/note_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from app.models.note import Note
from app.schemas.note import NoteCreate, NoteUpdate


class NoteService:
    def __init__(self, db: Session):
        self.db = db

    def create(self, user_id: str, data: NoteCreate) -> Note:
        note = Note(
            title=data.title,
            content=data.content,
            is_markdown=data.is_markdown,
            user_id=user_id,
            subject_id=data.subject_id,
            chapter_id=data.chapter_id,
        )
        self.db.add(note)
        self.db.commit()
        self.db.refresh(note)
        return note

    def get_all(self, user_id: str, subject_id: str = None, page: int = 1, limit: int = 20) -> tuple[list[Note], int]:
        query = self.db.query(Note).filter(Note.user_id == user_id)
        if subject_id:
            query = query.filter(Note.subject_id == subject_id)
        total = query.count()
        offset = (page - 1) * limit
        notes = query.order_by(Note.updated_at.desc()).offset(offset).limit(limit).all()
        return notes, total

    def get_by_id(self, user_id: str, note_id: str) -> Note:
        note = (
            self.db.query(Note)
            .filter(Note.id == note_id, Note.user_id == user_id)
            .first()
        )
        if not note:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Note not found",
            )
        return note

    def update(self, user_id: str, note_id: str, data: NoteUpdate) -> Note:
        note = self.get_by_id(user_id, note_id)
        update_data = data.model_dump(exclude_unset=True)
        for key, value in update_data.items():
            setattr(note, key, value)
        self.db.commit()
        self.db.refresh(note)
        return note

    def delete(self, user_id: str, note_id: str) -> None:
        note = self.get_by_id(user_id, note_id)
        self.db.delete(note)
        self.db.commit()
